platform/mbed_critical.h

```diff
@@ -124,7 +125,11 @@ bool core_util_atomic_flag_test_and_set(volatile core_util_atomic_flag *flagPtr)
  *
  * @param  flagPtr Target flag being cleared.
  */
-void core_util_atomic_flag_clear(volatile core_util_atomic_flag *flagPtr);
+MBED_FORCEINLINE void core_util_atomic_flag_clear(volatile core_util_atomic_flag *flagPtr)
+{
+    MBED_BARRIER();
+    flagPtr->_flag = false;
```

Comment:
You're right, one is needed there. I was thinking it wasn't needed for single-CPU, but sequential consistency requires it for the same reason you would need two DMBs (https://www.cl.cam.ac.uk/~pes20/cpp/cpp0xmappings.html). Assuming we want sequential consistency (the C++11 default), then we need to guarantee that

    atomic_store(x); // release
    atomic_load(y); // acquire

can't be reordered. So that has to end up as:

    BARRIER // for release semantics - keep preceding above
    x = foo;
    BARRIER // stop store+load being reordered
    bar = y;
    BARRIER // acquire semantics - keep following below

And it's store that gets the extra sequential-consistency barrier (like the extra DMB), because stores are probably less common.